internal/groups/api/endpoint.go

```diff
@@ -0,0 +1,316 @@
+		}
+		return buildGroupsResponse(page), nil
+	}
+}
+
+func ListGroupChildrenEndpoint(svc groups.Service) endpoint.Endpoint {
+	return func(ctx context.Context, request interface{}) (interface{}, error) {
+		req := request.(listGroupsReq)
+		if err := req.validate(); err != nil {
+			return groupPageRes{}, errors.Wrap(groups.ErrMalformedEntity, err)
+		}
+
+		page, err := svc.ListChildren(ctx, req.token, req.groupID, req.level, req.metadata)
+		if err != nil {
+			return groupPageRes{}, errors.Wrap(groups.ErrFetchGroups, err)
+		}
+
+		if req.tree {
+			return buildGroupsResponseTree(page), nil
+		}
+		return buildGroupsResponse(page), nil
+	}
+}
+
+func ListGroupParentsEndpoint(svc groups.Service) endpoint.Endpoint {
+	return func(ctx context.Context, request interface{}) (interface{}, error) {
+		req := request.(listGroupsReq)
+		if err := req.validate(); err != nil {
+			return groupPageRes{}, errors.Wrap(groups.ErrMalformedEntity, err)
+		}
+		page, err := svc.ListParents(ctx, req.token, req.groupID, req.level, req.metadata)
+		if err != nil {
+			return groupPageRes{}, errors.Wrap(groups.ErrFetchGroups, err)
+		}
+		if req.tree {
+			return buildGroupsResponseTree(page), nil
+		}
+		return buildGroupsResponse(page), nil
+
+	}
+}
+
+func AssignEndpoint(svc groups.Service) endpoint.Endpoint {
+	return func(ctx context.Context, request interface{}) (interface{}, error) {
+		req := request.(memberGroupReq)
+		if err := req.validate(); err != nil {
+			return nil, errors.Wrap(groups.ErrMalformedEntity, err)
+		}
+		if err := svc.Assign(ctx, req.token, req.memberID, req.groupID); err != nil {
+			return nil, errors.Wrap(groups.ErrAssignToGroup, err)
+		}
+		return assignMemberToGroupRes{}, nil
+	}
+}
+
+func UnassignEndpoint(svc groups.Service) endpoint.Endpoint {
+	return func(ctx context.Context, request interface{}) (interface{}, error) {
+		req := request.(memberGroupReq)
+		if err := req.validate(); err != nil {
+			return nil, errors.Wrap(groups.ErrMalformedEntity, err)
+		}
+		if err := svc.Unassign(ctx, req.token, req.memberID, req.groupID); err != nil {
+			return nil, errors.Wrap(groups.ErrUnassignFromGroup, err)
+		}
+		return removeMemberFromGroupRes{}, nil
+	}
+}
+
+func ListMembersEndpoint(svc groups.Service) endpoint.Endpoint {
+	return func(ctx context.Context, request interface{}) (interface{}, error) {
+		req := request.(listMemberGroupReq)
+		if err := req.validate(); err != nil {
+			return memberPageRes{}, errors.Wrap(groups.ErrMalformedEntity, err)
+		}
+		page, err := svc.ListMembers(ctx, req.token, req.groupID, req.offset, req.limit, req.metadata)
+		if err != nil {
+			return memberPageRes{}, err
+		}
+		return buildUsersResponse(page), nil
+	}
+}
+
+func buildGroupsResponseTree(page groups.GroupPage) groupPageRes {
+	groupsMap := map[string]*groups.Group{}
+	// Parents map keeps its array of children.
+	parentsMap := map[string][]*groups.Group{}
+	for _, g := range page.Groups {
+		group := g
```

Comment:
Why is this `group` var needed? Can't just `g` be used here directly?

Reply:
no it cant, it creates new object as a copy of `g` so that its address can be used for pointer in next lines

Reply:
Probably this would solve your issue:
```
	for i := range page.Groups {
		if _, ok := groupsMap[group.ID]; !ok {
			groupsMap[group.ID] = &page.Groups[i]
			parentsMap[group.ID] = make([]*groups.Group, 0)
		}
	}
```